rfid/chafon_driver.c

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Driver for Chafon RFID tags reader
+ * Copyright (c) 2024
+ * Mateusz Urbańczyk <mateusz.urbanczyk97@gmail.com>
+ */
+#include <linux/file.h>
+#include <linux/fs.h>
+#include <linux/kernel.h>
+#include <linux/miscdevice.h>
+#include <linux/module.h>
+#include <linux/mod_devicetable.h>
+#include <linux/of.h>
+#include <linux/of_address.h>
+#include <linux/platform_device.h>
+#include <linux/slab.h>
+#include <linux/uaccess.h>
+
+#include "reader.h"
+
+#define READ_BUFFER_SIZE 512
+
+enum ioctl_cmd {
+	IOCTL_SET_FILE = _IO('d', 0x01),
+	IOCTL_SET_SCAN_TIME = _IO('d', 0x02),
+	IOCTL_SET_POWER = _IO('d', 0x03),
+	IOCTL_SET_BUZZER = _IO('d', 0x04)
+};
+
+static int set_driver_file(struct rfid_reader *priv, int fd)
+{
+	if (priv->dev_fd.file)
+		fdput(priv->dev_fd);
+
+	priv->dev_fd = fdget(fd);
+
+	if (!priv->dev_fd.file)
+		return -EBADF;
+
+	return 0;
+}
+
+static long rfid_reader_ioctl(struct file *file,
+			      unsigned int cmd,
+			      unsigned long user_arg)
+{
+	struct rfid_reader *priv;
+	int kernel_arg;
+	int ret;
+
+	priv = container_of(file->private_data, struct rfid_reader, misc);
+	dev_notice(priv->dev, "[%s]: IOCTL cmd=%u, arg=%lu\n", __func__, cmd, user_arg);
+	ret = copy_from_user(&kernel_arg, (int __user *)user_arg, sizeof(int));
+
+	if (ret) {
+		dev_err(priv->dev, "[%s] ioctl failed copying from user: %d\n", __func__, ret);
+		return -EFAULT;
+	}
+
+	switch (cmd) {
+	case IOCTL_SET_FILE:
+		ret = set_driver_file(priv, kernel_arg);
+		break;
+	case IOCTL_SET_SCAN_TIME:
+		ret = set_scan_time(priv, priv->dev_fd.file, (uint8_t)kernel_arg);
+		break;
+	case IOCTL_SET_POWER:
+		ret = set_power(priv, priv->dev_fd.file, (uint8_t)kernel_arg);
+		break;
+	case IOCTL_SET_BUZZER:
+		ret = set_buzzer(priv, priv->dev_fd.file, (uint8_t)kernel_arg);
+		break;
+	default:
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static ssize_t rfid_reader_read(struct file *file,
+				char __user *user_buff,
+				size_t size,
+				loff_t *offset)
+{
+	char kernel_buffer[READ_BUFFER_SIZE];
+	struct rfid_reader *priv;
+	ssize_t num_bytes_read;
+
+	priv = container_of(file->private_data, struct rfid_reader, misc);
+
+	if (!priv->dev_fd.file)
+		return -ENODEV;
+
+	if (size > READ_BUFFER_SIZE) {
+		dev_notice(priv->dev, "[%s]: reading too much: %d\n", __func__, size);
+		return -ENOMEM;
+	}
+	read_tags(priv, priv->dev_fd.file);
+	num_bytes_read = 0;
+
+	if (num_bytes_read < 0)
+		return num_bytes_read;
+
+	if (copy_to_user(user_buff, kernel_buffer, num_bytes_read))
```

Comment:
`num_bytes_read` is 0 as it's assigned in line 99 and it's not changed anywhere

Reply:
please see: https://github.com/Tomatosoup97/chafon-rfid-driver/pull/1/files#diff-a7ed057690a580d5e3de4347a50f1fe70529ee68f2f69bd74b0446adc2000b0aR202-R221